### ISSUE TITLE
Redirect If Module Not Found

### ISSUE THREAD

**Issue by marwan-at-work** [gomods/athens#1167]
We have a global endpoint, but it's tied to the filter. I'm not too familiar with the filter mechanism, but we should potentially look into "redirecting" if a module does not exist. 

How to do that should be discussed here first before implementing. We can do this in the download.Protocol's main implementation, or it can be a middleware, or it can also be a wrapper implementation of the DP 

The original use case was from [this slack message](https://gophers.slack.com/archives/C9LRAQN8N/p1553533483794400)

```
I have an interesting problem I'd like to share:

I'm running Athens on a serverless platform that scales based on number of concurrent requests. The problem with that is that when the Athens' storage is all populated, the number of requests that determines scalability _should be high_ because all we do is just proxy the data. In other words, if a `go build` tries to download 100 modules concurrently, athens can do that in just one instance, without having to scale horizontally.

However, when the Athens' storage is empty, the number of requests that determines scalability _should be very low_ because each request corresponds to a `git clone`. And so if we do 100 git clones on one instance, `go build` will start getting 500 error codes. So in this case, we should probably tune our scaling threshold to be `3` instead of `100` which is a big difference.

I have three solutions:

1. just have Athens scale super quickly and call it a day (potentially have to pay more money and it doesn't make sense because we don't "stash" that often)

2. Change cloud platforms :smile: -- and have scalability be dependent on disk/memory/network

3. Have two Athens "services" on said platform...one that scales super quickly, and one that doesn't scale quickly. The one that does _not_ scale quickly, should be able to redirect to the one that scales quickly _only_ when a module is not found.

The last solution, is basically what we want to do (or already do?) for the GlobalEndpoint thingy that should probably renamed to just be like "redirect to multiple proxies when mod is not found"
```

Therefore this solution would be for option number 3 where one Athens instance scales slowly and redirects to another Athens instance that scales quickly. The one that scales slowly is _read only_ and never does any module-stashing.

**Comment by marpio** [comment]
@marwan-at-work 
Do you see this useful in other use cases beside supporting the serverless platform?

**Comment by marwan-at-work** [comment]
@marpio 
Yes, I imagine you might want to chain multiple proxies for various reasons. For example, an unauthenticated proxy that sits behind an authenticated proxy. Therefore the authenticated proxy only stores the private modules and not the public ones. 

That said, I do think we should probably re-think the GlobalEndpoint because it was originally designed for a totally different purpose (back in the Olympus days)

**Comment by arschles** [comment]
@marwan-at-work for option 3 above, how do you anticipate the redirecting tier filling its storage?

Also, I agree that we should re-think the GlobalEndpoint. Originally it was designed for mirroring - Athens would fill its storage from an upstream proxy (rather than a VCS). I still like that idea 😄

**Comment by marwan-at-work** [comment]
it can be done in a few ways: 

1. A download.Protocol wrapper
  1A: we could return a "RedirectError" and have the handler do the redirect part. 
  1B: we could just proxy the response from the redirect itself, which would still save us a lot of resource power because we're now just copying bytes over and not doing a git-clone. 
2. We can implement a download.Protocol middleware that has a different signature from the download.Protocol so it would be able to handle redirects appropriately and call next if it wishes to. 

I'm not exactly sure how 2 would pan out but it's worth looking into. At the end of the day, the handler needs to check the storage, and issue a 3xx if something is not found. Our code is pretty nicely separated and this type of logic couples the download.Protocol with the handlers themselves so we should definitely do it in a way that makes the most sense architecture-wise.

**Comment by arschles** [comment]
1(b) looks really good to me, it's almost exactly what I thought you were originally suggested in fact. One addition to that (we discussed this back in the early days as well) is to proxy the response from the redirect and copy it into storage (asynchronously, maybe, but that comes with some tradeoffs) - which makes the proxyer into a mirror
